frontend/lib/resolution/try-catch-analysis.cpp

```diff
@@ -0,0 +1,364 @@
+    if (!canThrow) {
+      context->error(ast, "cannot throw in a non-throwing function");
+    }
+    return true;
+  }
+  void TryCatchAnalyzer::exit(const Throw* ast, RV& rv) {
+  }
+
+
+  bool TryCatchAnalyzer::enter(const AstNode* ast, RV& rv) {
+    return true;
+  }
+  void TryCatchAnalyzer::exit(const AstNode* ast, RV& rv) {
+  }
+
+  void TryCatchAnalyzer::process(const uast::AstNode* symbol,
+                                   ResolutionResultByPostorderID& byPostorder) {
+    ResolvedVisitor<TryCatchAnalyzer> rv(context, symbol, *this, byPostorder);
+    // Traverse formals and then the body. This is done here rather
+    // than in enter(Function) because nested functions will have
+    // 'process' called on them separately.
+    if (auto fn = symbol->toFunction()) {
+      currentFn = fn;
+      if (fn->throws() && fn->name() == UniqueString::get(context,"deinit")) {
+        context->error(fn, "deinit is not permitted to throw");
+      }
+      if (auto body = fn->body()) {
+        // traverse the formals
+        for (auto formal : fn->formals()) {
+          formal->traverse(rv);
+        }
+        // traverse the real body
+        body->traverse(rv);
+      }
+    } else if (auto mod = symbol->toModule()) {
+      // module, comment, typedecl do same thing
+      // probably need a helper/adapter that we can re-use for this purpose
+      for (auto stmt : mod->stmts()) {
+        stmt->traverse(rv);
+      }
+    } else if (symbol->isTypeDecl()) {
+      // may want to specialize this in the future
+      symbol->traverse(rv);
+    } else {
+      symbol->traverse(rv);
+    }
+  }
+
+  // Compute the error handling mode to use by looking for modules whose kind is
+  // prototype or implicit. If neither of those apply, look for the value of the
+  // PERMIT_UNHANDLED_MODULE_ERRORS compiler flag to determine the handling mode
+  static ErrorCheckingMode computeErrorCheckingMode(Context* context,
+                                                    const AstNode* node) {
+    auto mode = ErrorCheckingMode::UNKNOWN;
+    auto moduleKind = parsing::idToModuleKind(context, node->id());
+    if (moduleKind == Module::Kind::PROTOTYPE ||
+        moduleKind == Module::Kind::IMPLICIT) {
+      mode = ErrorCheckingMode::FATAL;
+    }
+
+    if (mode == ErrorCheckingMode::UNKNOWN) {
+      // No mode was chosen explicitly, see if the compiler flag sets the
+      // handling mode or use the default.
+      if (isCompilerFlagSet(context,
+                            CompilerFlags::PERMIT_UNHANDLED_MODULE_ERRORS)) {
+        mode = ErrorCheckingMode::FATAL;
+      } else {
+        mode = ErrorCheckingMode::RELAXED;
+      }
+    }
+    CHPL_ASSERT(mode != ErrorCheckingMode::UNKNOWN);
+    return mode;
+  }
+
+  void checkThrows(Context* context, ResolutionResultByPostorderID& result,
+                   const AstNode* symbol) {
+    if (symbol->isFunction() || symbol->isModule()) {
```

Comment:
You have cases in `process` that support `TypeDecl` and maybe others, but how can they be hit if you only call process for functions and modules?